Add RecorderEventNotifier class

MichaelOrlov opened the pull request:
## Description

Add recorder event notifier class, which will be responsible for sending events on predefined topics about bag split and statistics about messages lost on transport and inner recorder level.

This PR is needed as part of the #2007 
>  2. Incremental statistics about the number of messages lost and bytes written on a per-topic basis should be published on a specific topic when messages are lost.

The `RecorderEventNotifier` class implemented as an extended replacement of the Messages Lost Notifier described in the 
[statistics_about_lost_and_recorded_messages.md](https://github.com/ros2/rosbag2/blob/7f304136062740cc718502dc4a855f79aa0571a3/docs/design/statistics_about_lost_and_recorded_messages.md?plain=1#L76C1-L108) design document.
```markdown
### 4. Messages Lost Notifier

   Will need to add a new `MessagesLostNotifier` class to the `rosbag2_transport::recorder` that 
   will publish on a predefined topic, incremental statistics about the number of lost messages 
   and bytes written on a per-topic basis when messages lost happened.
   - To facilitate this, the Rosbag2 recorder shall satisfy the following requirements:
      1. The messages lost event shall be published from the Rosbag2 recorder on a dedicated topic.
      e.g., `/events/rosbag2_messages_lost`
      2. The messages lost event shall include per-topic statistics with the number of lost 
        messages. The message type could be named as `rosbag2_interfaces::msg::MessagesLostEvent`
        and defined in the `rosbag2_interfaces` package. The message type will be similar to the 
        `rosbag2_interfaces::msg::WriteSplitEvent` message type and shall contain the following 
         fields:
         - `node_name` (string): The name of the node that is recording.
         - Array of the following fields:
         - `topic_name` (string): The name of the topic on which the messages were lost.
         - `messages_lost_in_transport` (uint64_t): The number of messages lost since the last 
           event on a DDS transport layer.
         - `messages_lost_in_recorder` (uint64_t): The number of messages lost since the last 
           event in the Rosbag2 recorder.
```

Besides the functionality for the message lost notification on a dedicated topic, the `RecorderEventNotifier` also encapsulates functionality for other event notifications, such as notifications about bag split. 
Notice that notifications about bag splits were built into the `RecorderImpl` class itself. Now that functionality was moved with minimum changes to the `RecorderEventNotifier` class. A queue was added for the bag_split events to avoid missing notifications when events happened one-to-one, with minimum delay.

**Note: In this PR, the `RecorderEventNotifier` doesn't implement the core functionality of the `Messages Lost Notifier` described in the [statistics_about_lost_and_recorded_messages.md](https://github.com/ros2/rosbag2/blob/rolling/docs/design/statistics_about_lost_and_recorded_messages.md) design document. However, it is a one-to-one replacement of the current implementation for the logic of what we have in the `RecorderImpl` class. That was made on purpose to be able to backport this PR to Kilted and Jazzy with minimum changes to reduce divergencies in the code base.
We will be able to backport the part related to the moving bag split event notification and callbacks for messages lost in the transport layer. However, the part related to the callbacks from the writer about lost messages will be omitted since the underlying API in the writer is missing.
The rest of the missing functionality for the `Messages Lost Notifier` will be implemented in follow-up PRs.**


<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.*
-->

Fixes # (issue) N/A
- Relates  #2007

### Is this user-facing behavior change?
No.

### Did you use Generative AI?
Yes, I am using GitHub Copilot, GPT-4.1 in my workflow.

### Additional Information
- Can be backported. Please refer to the notes above.
- It was considered to make one universal `EventNotifier` class for Player and Recorder. However, this design option was not accepted because events are very specific to the Player and Recorder classes, and it will be a cleaner implementation if they are split into separate classes. Having a common `EventNotifierBase` class also will not give much benefits, but rather will introduce extra layers of dependencies, which is unnecessary in this particular case.

MichaelOrlov committed:
Add RecorderEventNotifier class. Initial implementation

Signed-off-by: Michael Orlov <morlovmr@gmail.com>

MichaelOrlov committed:
Use RecorderEventNotifier in Recorder instead of event publisher thread

Signed-off-by: Michael Orlov <morlovmr@gmail.com>

MichaelOrlov committed:
Use PIMPL design pattern for the RecorderEventNotifier

Signed-off-by: Michael Orlov <morlovmr@gmail.com>

MichaelOrlov committed:
Add set_messages_lost_statistics_max_publishing_rate(hz) method

Signed-off-by: Michael Orlov <morlovmr@gmail.com>

MichaelOrlov committed:
Use queue for bag split events to avoid missing events

Signed-off-by: Michael Orlov <michael.orlov@apex.ai>

MichaelOrlov committed:
Disable publishing statistics by default

Signed-off-by: Michael Orlov <michael.orlov@apex.ai>

MichaelOrlov committed:
Add unit tests for the RecorderEventNotifier class

Signed-off-by: Michael Orlov <michael.orlov@apex.ai>

MichaelOrlov commented:
@christophebedard, @fujitatomoya I would appreciate your review here.

MichaelOrlov commented:
Pulls: ros2/rosbag2#2144
Gist: https://gist.githubusercontent.com/MichaelOrlov/0023bd074c23476d281e860b019ce139/raw/ba99cf9cffeb1510530a782cac7cd3e09d9e6263/ros2.repos
BUILD args:  --packages-above-and-dependencies rosbag2_transport rosbag2_tests
TEST args:  --packages-above rosbag2_transport rosbag2_tests
ROS Distro: rolling
Job: ci_launcher
ci_launcher ran: https://ci.ros2.org/job/ci_launcher/16752
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=24653)](http://ci.ros2.org/job/ci_linux/24653/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=18678)](http://ci.ros2.org/job/ci_linux-aarch64/18678/)
* Linux-rhel [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-rhel&build=4103)](http://ci.ros2.org/job/ci_linux-rhel/4103/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=25175)](http://ci.ros2.org/job/ci_windows/25175/)

MichaelOrlov committed:
Address Windows build warnings in tests

- Capture 'iterations_per_thread' variable explicitly in lambda

Signed-off-by: Michael Orlov <michael.orlov@apex.ai>

MichaelOrlov commented:
Re-run Windows CI job after attempt to fix build errors.
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=25176)](http://ci.ros2.org/job/ci_windows/25176/)

MichaelOrlov committed:
Use uint64_t for counters of the total messages lost

- Reasoning. The consumer of the counters i.e. 'event/messages_lost'
  message type also has uint64_t types for values.
  Anyway it is more safer to use uint64_t type explicitly for the
  accumulating data.

Signed-off-by: Michael Orlov <michael.orlov@apex.ai>

kylemarcey commented on rosbag2_transport/src/rosbag2_transport/recorder_event_notifier.cpp:
```diff
@@ -0,0 +1,80 @@
+
+RecorderEventNotifier::~RecorderEventNotifier()
+{
+  // Explicitly reset the pimpl_ to ensure the destructor of RecorderEventNotifierImpl is called
+  // only once.
+  pimpl_.reset();
```
@MichaelOrlov why do you need to explicitly reset this instead of just letting the unique_ptr dtor do it automatically?

MichaelOrlov replied:
Yeah, it was a surprise for me too.
The simple fist test
```cpp
TEST_F(TestRecorderEventNotifier, default_ctor_dtor)
{
  auto node = std::make_unique<rclcpp::Node>("test_recorder_event_notifier");
  auto notifier = std::make_unique<RecorderEventNotifier>(node_.get());
  // Destructor should not throw
  ASSERT_NO_THROW(notifier.reset());
}
```
Was catching a segmentation fault when exiting from test. i.e. when `notifier` was trying to be destructed since it goes out of scope.
I've tried to debug this phenomenon and it turns out that after the first destruction call via 
```cpp
  ASSERT_NO_THROW(notifier.reset());
```
The inner `pimpl_` is not `nullptr` and still points to the same memory address as before the first destruction call.
And we got a double destruction of the same object.

kylemarcey commented on rosbag2_transport/src/rosbag2_transport/recorder_event_notifier_impl.hpp:
```diff
@@ -0,0 +1,236 @@
+        RCLCPP_DEBUG(node->get_logger(), "Messages lost statistics publishing is disabled");
+      } else if (update_rate_hz > 0.0f) {
+        disable_publishing_msgs_lost_statistics_ = false;
+        msgs_lost_stats_max_publishing_period_ =
+          std::chrono::milliseconds(static_cast<int>(1000 / update_rate_hz));
```
@MichaelOrlov is there any validation of `update_rate_hz`? If it is set (probably mistakenly) to more than 1000, the period will be 0 and I think you will busy loop.

MichaelOrlov replied:
@kylemarcey Good catch I will add an extra check and throw an exception

MichaelOrlov committed:
Throw exception if update_rate_hz >= 1000.0f

- Reasoning: To avoid a busy loop when condition_variable timeout would
 be zero.

Signed-off-by: Michael Orlov <michael.orlov@apex.ai>

kylemarcey commented:
@MichaelOrlov ok, LGTM

MichaelOrlov commented:
@christophebedard @fujitatomoya This PR has already been reviewed. Can some of you additionally approve it to be able to merge it?

MichaelOrlov commented:
Pulls: ros2/rosbag2#2144
Gist: https://gist.githubusercontent.com/MichaelOrlov/a9f3377b3931536f159c2f7d1f258fef/raw/ba99cf9cffeb1510530a782cac7cd3e09d9e6263/ros2.repos
BUILD args:  --packages-above-and-dependencies rosbag2_transport rosbag2_tests
TEST args:  --packages-above rosbag2_transport rosbag2_tests
ROS Distro: rolling
Job: ci_launcher
ci_launcher ran: https://ci.ros2.org/job/ci_launcher/16768
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=24673)](http://ci.ros2.org/job/ci_linux/24673/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=18699)](http://ci.ros2.org/job/ci_linux-aarch64/18699/)
* Linux-rhel [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-rhel&build=4121)](http://ci.ros2.org/job/ci_linux-rhel/4121/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=25194)](http://ci.ros2.org/job/ci_windows/25194/)

MichaelOrlov commented:
The failed Linux CI job was auto-restarted due to the CI infrastructure failure and passed green
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=24675)](http://ci.ros2.org/job/ci_linux/24675/)

MichaelOrlov commented:
https://github.com/Mergifyio backport kilted
